### PR TITLE
feat(layout/#592): layout tabs - part 1

### PR DESCRIPTION
**Pull request by glennsl**
layout tabs

This adds a new concept of "layout tabs", which behave more like the tabs in vim. By default these tabs are positioned at the bottom to avoid two levels of tabs at the top, but they can be positioned at the top using the `oni.layout.layoutTabPosition` configuration setting. They can also be hidden, or made to  always show, using `oni.layout.showLayoutTabs`.

To mimic the layout of vim, use:

```
"oni.layout.layoutTabPosition": "top",
"editor.workbench.showTabs": false,
```

There's still a bit of work to get faithful vim behaviour though. These are the next steps I plan to work on:

- [ ] Wire up the `:tab*` commands (`:tabnew` and `:tabedit` already work, although not entirely correctly)
- [ ] Wire up `gt` and `gT`. This must be done from `libvim` for the moment due to bugs in `editor-input`, but shouldn't be too hard.
- [ ] Add a configuration setting for a "single tab mode" that will reuse the first tab instead fo adding hidden tabs when `showTabs` is off.
- [ ] For technical reasons it is currently not possible to open a new buffer directly in a new split or layout tab. Instead what will happen is that the new split or layout tab is opened with he currently active editor in addition to the new one. This is because opening a buffer will automatically open an editor (via `BufferEnter`), and it's not possible to open an empty split or layout tab before opening the buffer. Therefore the only two options are to either open a new split or layout tab with the currently active editor and then open the buffer, or to open the buffer in the old split/layout tab before opening a new one. Either way there'll be an extra editor that shouldn't be there.

Addresses #592

### CODE REVIEW

**Comment by bryphe** [comment]
Wow, this is looking great @glennsl ! Thanks for the work on this.

**Comment by bryphe** [comment]
>  Wire up gt and gT. This must be done from libvim for the moment due to bugs in editor-input, but shouldn't be too hard.

I can help with this piece, if you'd like, and haven't started it -

**Comment by bryphe** [comment]
One thing to think about that isn't called out on your list is the `quit` behavior, too - if `:q` is used in the layout tab, it should be closed, moving to the previous tab. If it's the last layout tab, the editor should be closed.

File: src/Components/Tabs.re [hunk 1]
```diff
@@ -49,7 +49,7 @@ let schedulePostRender = f => postRenderQueue := [f, ...postRenderQueue^];
 let component = React.Expert.component("Tabs");
 let make =
     (
-      ~children as render: 'a => element,
+      ~children as render: (~isSelected: bool, int, 'a) => element,
```
Comment (bryphe) [inline src/Components/Tabs.re:52]: I wonder if this `int` should be a named parameter as well, like `~index:int`?

File: src/Feature/Layout/Model.re [hunk 2]
```diff
@@ -132,57 +132,80 @@ type model = {
...
+    }
+  );
+
+let windows = model => Layout.windows(model |> activeLayout |> activeTree);
 
 let visibleEditors = model =>
   model
   |> windows
-  |> List.filter_map(id => groupById(id, model))
+  |> List.filter_map(id => model |> activeLayout |> groupById(id))
```
Comment (bryphe) [inline src/Feature/Layout/Model.re:204]: Nice! Seems like adding the layout tabs was a pretty straightforward addition to the model, with the groundwork you laid out in #1977

File: src/Feature/Layout/View.re [hunk 3]
```diff
@@ -0,0 +1,616 @@
...
+
+  let id: t => int;
+  let title: t => string;
+  let icon: t => option(Oni_Core.IconTheme.IconDefinition.t);
+  let isModified: t => bool;
+
+  let render: (~isActive: bool, t) => Revery.UI.element;
+};
+
+module Tab = {
```
Comment (bryphe) [inline src/Feature/Layout/View.re:26]: Thanks for generalizing this

**Comment by glennsl** [comment]
> I can help with this piece, if you'd like, and haven't started it -

Already working on it. Implemented it twice even! First as an extension of the goto callback, but then after discovering it was tightly (and weirdly) tied to multiple `:tab` commands I decided to go with a separate tabpage callback.

> One thing to think about that isn't called out on your list is the quit behavior, too - if :q is used in the layout tab, it should be closed, moving to the previous tab. If it's the last layout tab, the editor should be closed.
>
> Related to this - if I close (press the x) of the active tab, when there are tabs available, the editor closes. I would expect it to switch to the next available tab.

Ah, yeah, sorry. That behaviour was intended only for when the last layout tab was closed. I forgot to test and handle these other cases, so it just falls back to that. Will fix!

**Comment by glennsl** [comment]
The Ubuntu 16 CI seems to fail consistently on `esy bootstrap` now. I think this is the relevant part of the log:

```
    autoconf.texi:8017: misplaced }
    autoconf.texi:8018: must be after `@defmac' to use `@defmacx'
    autoconf.texi:8019: misplaced }
    autoconf.texi:8206: must be after `@defmac' to use `@defmacx'
    autoconf.texi:8271: misplaced }
    autoconf.texi:8290: misplaced }
    autoconf.texi:8317: misplaced }
    autoconf.texi:8380: must be after `@defmac' to use `@defmacx'
    conftest.c:4597: must be after `@defmac' to use `@defmacx'
    conftest.c:15929: must be after `@defmac' to use `@defmacx'
    Makefile:251: recipe for target 'autoconf.info' failed
```

Makes no sense to me though...

**Comment by bryphe** [comment]
> Already working on it. Implemented it twice even! First as an extension of the goto callback, but then after discovering it was tightly (and weirdly) tied to multiple :tab commands I decided to go with a separate tabpage callback.

Cool, sounds good. Thanks for looking into it!

> The Ubuntu 16 CI seems to fail consistently on esy bootstrap now. I think this is the relevant part of the log:

Ouch, ya, I'm looking into it - it's unrelated to your changes. I also see this in the log:
```
    /home/vsts/.esy/3/b/autoconf-dafb3f25/build-aux/missing: line 51: aclocal-1.11: command not found
    WARNING: 'aclocal-1.11' is missing on your system.  You should only need it if
             you modified 'acinclude.m4' or 'configure.ac'.  You might want
             to install the Automake and Perl packages.  Grab them from
             any GNU archive site.
```

It shouldn't need to run `aclocal` at all, so it's curious why it's trying... In the meantime, if this is the only check that's failing - feel free to ignore / override. I'll work on getting that check back up.

File: src/Components/Tabs.re (hunk 1 above)
Comment (glennsl) [reply src/Components/Tabs.re]: Fixed in https://github.com/onivim/oni2/pull/2013/commits/f43e24dfdc0e5346d08bdf825c2c8143fa3979e3

**Comment by glennsl** [comment]
The closing behaviour should be fixed now. But please do check thoroughly, as it's surprisingly complex with all the different levels and ways of closing.

**Pull request merged**